keps/sig-cli/20201216-kubectl-default-container.md

```diff
@@ -0,0 +1,147 @@
+---
+title: Default Container for Kubectl logs and exec
+authors:
+  - "@mfojtik"
+  - "@pacoxu"
+owning-sig: sig-cli
+participating-sigs:
+reviewers:
+  - "@dougsland"
+  - "@eddiezane"
+  - "@soltysh"
+
+approvers:
+  - ""
+
+editor: TBD
+creation-date: 2020-12-16
+last-updated: 2020-12-16
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# default container behavior
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Many Kubectl commands make requests to specific containers, `kubectl logs` and `kubectl exec` are most commonly
+used.
+
+All commands that need to specify container name:
+
+- kubectl attach - Attach to a running container
+- kubectl cp - Copy files and directories to and from containers.
+- kubectl exec - Execute a command in a container
+
+These three commands above are similar. If omitted, the first container in the pod will be chosen.
+
+
+- kubectl logs - Print the logs for a container in a pod
+
+`kubectl logs` is a little different. Not default value for it. 
+However it support to logs all containers with `--all-containers`.
+
+https://github.com/kubernetes/kubernetes/pull/87809 already add an annotation named
+`kubectl.kubernetes.io/default-logs-container` for log to use as default.
+
+We don't have a general default container name attribute for pod, and this would change pod spec and not acceptable.
+Why not use `kubectl.kubernetes.io/default-container` as the default container name annotation for all four commands
```

Comment:
I believe we could start with a single and generic annotation for all above commands.

Reply:
then I will update the kep as well as the PR. Need I make it compatible with `kubectl.kubernetes.io/default-logs-container` for `kubectl log` or not?